python/contracts.py

```diff
@@ -0,0 +1,320 @@
+
+    def __call__(self, v):
+        try:
+            return self.func(v)
+        except:
+            # An exception implies the requirement is not met
+            return False
+
+
+class Type(Requirement):
+    """
+    Type: a `Requirement` that checks argument types (maybe 1d-nested). For example:
+          Type(0) requires the argument to be an integer,
+          Type([Int&Positive]) requires the argument to be a list of positive integers.
+    """
+    def __init__(self, val):
+        def deepcheck(v):
+            # NOTE: 1. list that's more than 2d degrades to list at the moment
+            #       2. only the 1st element of `val` is respected as the requirements
+            #          to all elements of `v`
+            # TODO(shendiaomo): support checking dict like `Type({String: Int})`?
+            if type(val) in (list, tuple) and len(val) and isinstance(
+                    val[0], Requirement):
+                return (type(v) in (list, tuple)
+                        and all(map(lambda i: val[0](i), v)))
+            return isinstance(v, type(val))
+
+        if type(val) in (list, tuple) and len(val) and isinstance(
+                val[0], Requirement):
+            typename = "list[{}]".format(val[0].diag)
+        else:
+            typename = type(val).__name__
+        super().__init__(deepcheck, "TYPE={}".format(typename))
+
+
+class Between(Requirement):
+    """
+    Between: a `Requirement` to the range of an argument. Bounds are included (closed interval).
+    """
+    def __init__(self, lower, upper):
+        super().__init__(lambda v: upper >= v >= lower,
+                         "BETWEEN ({},{})".format(lower, upper))
+
+
+class Greater(Requirement):
+    """
+    Greater: a `Requirement` to the lower bound of an argument. Synonym of the operator `>`.
+    """
+    def __init__(self, val):
+        super().__init__(lambda v: v > val, "'>{}'".format(val))
+
+
+class GreaterEqual(Requirement):
+    """
+    GreaterEqual: a `Requirement` to the lower bound of an argument. Synonym of the operator `>=`.
+    """
+    def __init__(self, val):
+        super().__init__(lambda v: v >= val, "'>={}'".format(val))
+
+
+class Less(Requirement):
+    """
+    Less: a `Requirement` to the upper bound of an argument. Synonym of the operator `<`.
+    """
+    def __init__(self, val):
+        super().__init__(lambda v: v < val, "'>{}'".format(val))
+
+
+class LessEqual(Requirement):
+    """
+    LessEqual: a `Requirement` to the upper bound of an argument. Synonym of the operator `<=`.
+    """
+    def __init__(self, val):
+        super().__init__(lambda v: v <= val, "'<={}'".format(val))
+
+
+class In(Requirement):
+    """
+    In: a `Requirement` to the value of an argument. Synonym of the keyword `in`.
+    """
+    def __init__(self, *args):
+        super().__init__(lambda v: v in args,
+                         "IN ({})".format(",".join(str(i) for i in args)))
+
```

Comment:
Maybe add `NotIn` ? Likely, `NotEmpty` for **string** and container types. Of course we can add them when required.